plugins/outputs/opensearch/opensearch.go

```diff
@@ -0,0 +1,368 @@
+	Log                 telegraf.Logger `toml:"-"`
+	pipelineName        string
+	pipelineTagKeys     []string
+	tagKeys             []string
+	httpconfig.HTTPClientConfig
+
+	client *elastic.Client
+}
+
+//go:embed telegrafTemplate.json
+var telegrafTemplate string
+
+type templatePart struct {
+	TemplatePattern string
+}
+
+func (*Opensearch) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *Opensearch) Connect() error {
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(a.Timeout))
+	defer cancel()
+
+	var clientOptions []elastic.ClientOptionFunc
+
+	ctxt := context.Background()
+	httpclient, err := a.HTTPClientConfig.CreateClient(ctxt, a.Log)
+	if err != nil {
+		return err
+	}
+
+	osURL, err := url.Parse(a.URLs[0])
+	if err != nil {
+		return fmt.Errorf("parsing URL failed: %v", err)
+	}
+
+	clientOptions = append(clientOptions,
+		elastic.SetHttpClient(httpclient),
+		elastic.SetSniff(a.EnableSniffer),
+		elastic.SetScheme(osURL.Scheme),
+		elastic.SetURL(a.URLs...),
+		elastic.SetHealthcheckInterval(time.Duration(a.HealthCheckInterval)),
+		elastic.SetHealthcheckTimeout(time.Duration(a.HealthCheckTimeout)),
+		elastic.SetGzip(a.EnableGzip),
+	)
+
+	if a.Username != "" && a.Password != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetBasicAuth(a.Username, a.Password),
+		)
+	}
+
+	if a.AuthBearerToken != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetHeaders(http.Header{
+				"Authorization": []string{fmt.Sprintf("Bearer %s", a.AuthBearerToken)},
+			}),
+		)
+	}
+
+	if time.Duration(a.HealthCheckInterval) == 0 {
+		clientOptions = append(clientOptions,
+			elastic.SetHealthcheck(false),
+		)
+		a.Log.Debugf("Disabling health check")
+	}
+
+	a.client, err = elastic.NewClient(clientOptions...)
+	if err != nil {
+		return err
+	}
+
+	if a.ManageTemplate {
+		err := a.manageTemplate(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getPointID generates a unique ID for a Metric Point
+func getPointID(m telegraf.Metric) string {
+	var buffer bytes.Buffer
+	//Timestamp(ns),measurement name and Series Hash for compute the final SHA256 based hash ID
+
+	buffer.WriteString(strconv.FormatInt(m.Time().Local().UnixNano(), 10)) //nolint:revive // from buffer.go: "err is always nil"
+	buffer.WriteString(m.Name())                                           //nolint:revive // from buffer.go: "err is always nil"
+	buffer.WriteString(strconv.FormatUint(m.HashID(), 10))                 //nolint:revive // from buffer.go: "err is always nil"
+
+	return fmt.Sprintf("%x", sha256.Sum256(buffer.Bytes()))
+}
+
+func (a *Opensearch) Write(metrics []telegraf.Metric) error {
+	bulkRequest := a.client.Bulk()
+
+	for _, metric := range metrics {
+		var name = metric.Name()
+
+		// index name has to be re-evaluated each time for telegraf
+		// to send the metric to the correct time-based index
+		indexName := a.GetIndexName(a.IndexName, metric.Time(), a.tagKeys, metric.Tags())
```

Comment:
```suggestion
		var indexName strings.Builder
		templateMetric := TemplateMetric{metric}

		// indexTemplate is of type text/template.Template
		err := r.indexTemplate.Execute(&indexName, &templateMetric)
```
And similar for pipeline, then the `GetReplacementKeys`, `GetIndexName` and `getPipelineName` methods can be removed..

Reply:
could you please provide any example in the existing plugins which I can refer to make the change you've proposed.

Reply:
The simplest is have a look at the template processor of course. But I mainly provided already the code needed.

Reply:
Dear Hipska, I'm facing some compilation errors like below-
`implicit assignment to unexported field metric in "github.com/influxdata/telegraf/plugins/processors/template".TemplateMetric `

Don't think the template is ready for this implementation, being a newbie to the GO World, would really appreciate if you can accept/merge the PR and make the change yourselves. I'll follow your approach and will take care of any subsequent issues/changes going forward.

Reply:
Can you please join us in the #telegraf-dev channel on Slack? (I sent you an invite) Then we can discuss and help you out..